### PR TITLE
Add defer, a general facility for ad-hoc RAII.

### PR DESCRIPTION
**Pull request by wilhuff**
Also, use it in a few places where it helps.

### CODE REVIEW

File: Firestore/core/src/util/defer.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * Storage for a deferred action. The `action` is invoked during the destructor
+ * of the `Deferred`.
+ */
+template <typename Action>
+class Deferred {
+ public:
+  explicit Deferred(Action&& action) : action_(std::move(action)) {
+  }
+
```
Comment (var-const) [inline Firestore/core/src/util/defer.h:60]: Should this class forbid copying (and probably moving as well)?

File: Firestore/core/src/util/defer.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+namespace firebase {
+namespace firestore {
+namespace util {
+
+template <typename Action>
+class Deferred;
+
+/**
+ * Creates a deferred action that will be destroyed at the close of the lexical
```
Comment (var-const) [inline Firestore/core/src/util/defer.h:30]: Looks like `destroyed` should be `executed`/`run`/`invoked`/etc.

File: Firestore/core/src/util/defer.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * scope containing the result of this function call. This is useful for
```
Comment (var-const) [inline Firestore/core/src/util/defer.h:31]: Optional: perhaps highlight that any meaningful use of this function requires assigning its return value to a variable?

File: Firestore/core/src/util/defer.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ *         fclose(file);
+ *       }
+ *     });
+ *
+ * @param action a callable object; usually a lambda. Even if exceptions are
+ *     enabled, when `action` is invoked it must not throw. This is similar to
+ *     the restriction that exists on destructors generally.
+ */
+template <typename Action>
+Deferred<Action> defer(Action&& action) {
```
Comment (var-const) [inline Firestore/core/src/util/defer.h:47]: Question: what's the rationale to use a lowercase name for this function? (generally, the style guide requires functions to be `CamelCase`)

File: Firestore/core/src/util/defer.h (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/src/util/defer.h]: Fixed.

File: Firestore/core/src/util/defer.h (hunk 3 above)
Comment (wilhuff) [reply Firestore/core/src/util/defer.h]: Done.

File: Firestore/core/src/util/defer.h (hunk 4 above)
Comment (wilhuff) [reply Firestore/core/src/util/defer.h]: It's an oversight. Fixed.

File: Firestore/core/src/util/defer.h (hunk 1 above)
Comment (wilhuff) [reply Firestore/core/src/util/defer.h:65]: In order to return this by value the type must at least be moveable. Only once we get to C++17 with guaranteed elision can we return a non-copyable, non-moveable object by value. I've made it move-only.

The other alternative is to make this class concrete, taking a `std::function<void()>` as an argument. Then the constructor can be invoked directly and the class could be made non-copyable and non-moveable (like `std::lock_guard`). Usage would look like this:

```c++
FILE* file = fopen(filename, "rb");
Defer cleanup([&] {
  if (file) {
    fclose(file);
  }
}
```

Unfortunately, the major downside of that approach is that `std::function` must perform a heap allocation whereas the inferred implementation defined type of `Action` here likely does not.

**Pull request merged**